### PR TITLE
Node group based rel table

### PR DESCRIPTION
**Pull request by ray6080**
This is a quite bulky PR, sorry for that :).
Mainly this PR introduces several major changes:
- Moved rel table to be stored as node groups.
- Remove arrow from third party dependencies. Now all reads are done through our customized CSV and Parquet readers.
- A bunch of clean ups, including removing legacy data structures for rel tables, such as Lists, Columns, etc., cleaning up code related to `StorageStructure` and a bunch of util functions for column and list files.

**Note that this PR breaks updates, creations, deletions for rel tables for a short period of time**
We discussed this internally inside the team, though it's not a good practice to break master, we decided to merge this early to save efforts for everyone to sync around monster changes brought by this PR.

### Rel Table Storage
A rel table consists of both forward and backward `RelTableData`, which consists of multiple `Columns`.
Each `Column` can be stored in two formats: `REGULAR` and `CSR`. Regular columns are used by both NodeTable and RelTable when the multiplicity is `ONE/MANY_TO_ONE`, thus, each node offset corresponds to one value in the column. CSR columns are used only by RelTable when the multiplicity is `ONE/MANY_TO_MANY`, where each node offset corresponds to a CSR list in the column.

For a node group of `RelTableData` stored in CSR format, it contains following column chunks:
- CSR offsets: stores for each node in the node group its csr offset. The column chunk contains NODE_GROUP_SIZE values, each of which records the end offset of the CSR of the node with the corresponding offset. For example, csr offsets are [3, 5, 5, 9, ... ], meaning the first CSR list's size is 3, and the second is 2, etc.
When reading data from a node group, we always need to first read out its csr offsets into `RelDataReadState`, then read actual data columns based on the offsets.
- Adj column: stores neighbour node ids. we should consider renaming this to `nbrColumn` later.
- Properties: rel properties defined in the table schema.

### COPY
COPY pipelines for rel tables are as follows:
```
            __RESULT_COLLECTOR__
           |           |       |
COPY_REL(fwd)   COPY_REL(bwd)  PARTITIONER     
                                    |
                              INDEX_LOOKUP
                                    |
                                 READER
```
The main part is executed as three pipelines.
The right most `PARTITIONER` pipeline gets executed first. It performs partitioning of all tuples based on their src and dst node offsets. We copy tuples to each partitioning buffer as for now and organize them into different partitions (i.e., node groups). This could contribute to large memory consumption when the copied dataset is large. The following solution to this is to allow spilling intermediate partitioning result to disk as tmp files, which introduces I/O overheads, but the I/O pattern should be always be sequential reads and writes.
Other two pipelines are both `CopyRel` for each directions, which consume partitioned buffer from the `PARTITIONER` pipeline. During execution, each thread will fetch one partition (i.e., node group) at a time, construct necessary column chunks for the node group, and append them to table data file.

Note that this design is not able to handle very skewed dataset well. When one node group is extremely large, the copy rel pipeline could end up being only one thread working during most of the time, we should perhaps address this by further partitioning the extremely large node group and let multiple threads work concurrently on it.

### TODOs
- [ ] Throw error messages in the front end for disabled updates/deletions/insertions on rel table.

**Tests**
- [ ] Add tests on nested data types for rel table. `STRUCT`, `MAP`, `UNION`.
- [ ] Fix e2e ddl tests.
- [ ] Fix RDF tests. `CopyRDFTest` and `RdfoxExample.Basic`.
- [ ] Fix `CopyRelTableMultiplicityViolationTest` and `TinySnbExceptionTest.DeleteNodeWithEdgeErrorTest`.
- [ ] Fix `TinySnbCreateNodeTest`, `TinySnbCreateReadRelTest`, `TinySnbDeleteRelTest`, `TinySnbMergeRelTest`, `TinySnbSetReadRelTest`, `UpdateRelTest`, `DeleteRelTest`, `CreateRelTest`, `TCK`, `DemoDBCreateTest`, `DemoDBDeleteTest`.

**Refactoring**
- [ ] Change `RelDataDirection` from `enum` to `enum class`.
- [ ] Add OFFSET logical data type #2331.
- [ ] Address this [comment](https://github.com/kuzudb/kuzu/pull/2246/files#r1374663171) on getting rid of the converting binding type of SERIAL to INT64.
- [ ] Address this [comment](https://github.com/kuzudb/kuzu/pull/2246/files#r1374714125) on refactoring copyDataToPartitions function.
- [ ] Revisit VAR_LIST storage layout. Existing one is not write friendly. Consider the alternative way similar to STRING (indices, offsets, data). I will open a separate issue on this.
- [ ] Move offset and adj column to properties.
- [ ] Get rid of `leftNumRows` inside `Reader::readNextDataChunk()`.
- [ ] Turn on compression on internalID.

### CODE REVIEW

File: src/processor/map/map_copy_from.cpp [hunk 1]
```diff
@@ -90,100 +78,43 @@ std::unique_ptr<PhysicalOperator> PlanMapper::mapCopyNodeFrom(LogicalOperator* l
...
-                direction, property->getPropertyID(), DBFileType::ORIGINAL);
-            relLists->propertyLists.emplace(property->getPropertyID(),
-                InMemListsFactory::getInMemPropertyLists(fName, *property->getDataType(), numNodes,
-                    csvReaderConfig, relLists->adjList->getListHeadersBuilder()));
-        }
-        directedInMemRelData->setRelLists(std::move(relLists));
-    }
-    return directedInMemRelData;
+    auto dataColumnPositions = getExpressionsDataPos(copyFromInfo->dataColumnsToCopy, *outFSchema);
+    auto dataFormat = tableSchema->isSingleMultiplicityInDirection(direction) ?
```
Comment (andyfengHKU) [inline src/processor/map/map_copy_from.cpp:109]: open an issue about removing `isSingleMultiplicityInDirection` for me

File: src/processor/map/map_copy_from.cpp [hunk 2]
```diff
@@ -192,25 +123,28 @@ std::unique_ptr<PhysicalOperator> PlanMapper::mapCopyRelFrom(
...
-        copyFromInfo->fileScanInfo->readerConfig->csvReaderConfig.get());
-    auto bwdRelData = initializeDirectedInMemRelData(RelDataDirection::BWD, tableSchema,
-        storageManager.getNodesStore(), storageManager.getDirectory(),
-        copyFromInfo->fileScanInfo->readerConfig->csvReaderConfig.get());
+    auto prevOperator = mapOperator(copyFrom->getChild(0).get());
+    assert(prevOperator->getOperatorType() == PhysicalOperatorType::PARTITIONER);
+    auto partitionerSharedState = dynamic_cast<Partitioner*>(prevOperator.get())->getSharedState();
+    partitionerSharedState->numPartitions.resize(2);
+    std::vector<std::unique_ptr<LogicalType>> columnTypes;
+    columnTypes.push_back(std::make_unique<LogicalType>(LogicalTypeID::INTERNAL_ID)); // ADJ COLUMN.
```
Comment (andyfengHKU) [inline src/processor/map/map_copy_from.cpp:131]: leave TODO for me to bind column types in binder.

File: src/processor/map/map_copy_from.cpp [hunk 3]
```diff
@@ -90,100 +78,43 @@ std::unique_ptr<PhysicalOperator> PlanMapper::mapCopyNodeFrom(LogicalOperator* l
...
-    } else {
-        assert(prevOperator->getChild(0)->getOperatorType() == PhysicalOperatorType::READER);
-        reader = reinterpret_cast<Reader*>(prevOperator->getChild(0));
-    }
-
-    auto tableSchema = reinterpret_cast<RelTableSchema*>(copyFromInfo->tableSchema);
-    auto internalIDDataPos =
+    auto tableSchema = dynamic_cast<RelTableSchema*>(copyFromInfo->tableSchema);
+    auto partitioningIdx = direction == FWD ? 0 : 1;
+    auto maxBoundNodeOffset =
```
Comment (andyfengHKU) [inline src/processor/map/map_copy_from.cpp:89]: shouldn't this be done during init global state?

**Review by acquamarin — APPROVED** [review]

File: src/include/processor/operator/partitioner.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    void initGlobalStateInternal(ExecutionContext* context) final;
+    void initLocalStateInternal(ResultSet* resultSet, ExecutionContext* context) final;
+    void executeInternal(ExecutionContext* context) final;
+
+    inline std::shared_ptr<PartitionerSharedState> getSharedState() { return sharedState; }
+
+    std::unique_ptr<PhysicalOperator> clone() final;
+
+private:
+    static void initializePartitioningStates(
```
Comment (andyfengHKU) [inline src/include/processor/operator/partitioner.h:77]: Do u need static functions? I thought member function is enough.

File: src/binder/bind/bind_copy.cpp [hunk 5]
```diff
@@ -118,21 +118,19 @@ std::unique_ptr<BoundStatement> Binder::bindCopyNodeFrom(
...
     return std::make_unique<BoundCopyFrom>(std::move(boundCopyFromInfo));
 }
 
 std::unique_ptr<BoundStatement> Binder::bindCopyRdfNodeFrom(
     std::unique_ptr<ReaderConfig> readerConfig, TableSchema* tableSchema) {
     auto containsSerial = bindContainsSerial(tableSchema);
     auto stringType = LogicalType{LogicalTypeID::STRING};
-    auto nodeID =
-        createVariable(std::string(Property::INTERNAL_ID_NAME), LogicalTypeID::INTERNAL_ID);
+    auto nodeID = createVariable(std::string(Property::INTERNAL_ID_NAME), LogicalTypeID::INT64);
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/binder/bind/bind_copy.cpp:133]: Good idea. I will do that in a separate PR after this one.

File: src/binder/bind/bind_copy.cpp [hunk 6]
```diff
@@ -158,32 +156,38 @@ std::unique_ptr<BoundStatement> Binder::bindCopyRelFrom(
...
-    auto columns = bindExpectedRelFileColumns(tableSchema, *readerConfig);
-    auto srcKey = columns[0];
-    auto dstKey = columns[1];
-    auto relID =
-        createVariable(std::string(Property::INTERNAL_ID_NAME), LogicalTypeID::INTERNAL_ID);
+    auto columnsToRead = bindExpectedRelFileColumns(tableSchema, *readerConfig);
+    auto relID = createVariable(std::string(Property::INTERNAL_ID_NAME), LogicalTypeID::INT64);
     auto boundFileScanInfo = std::make_unique<BoundFileScanInfo>(
-        std::move(readerConfig), std::move(columns), std::move(relID), TableType::REL);
+        std::move(readerConfig), columnsToRead, relID->copy(), TableType::REL);
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/binder/bind/bind_copy.cpp:162]: I actually need `relID` at L177, where it is moved.

File: src/binder/bind/bind_copy.cpp [hunk 7]
```diff
@@ -158,32 +156,38 @@ std::unique_ptr<BoundStatement> Binder::bindCopyRelFrom(
...
+    auto dstNodeID =
+        createVariable(std::string(Property::REL_NBR_OFFSET_NAME), LogicalTypeID::INT64);
     auto extraCopyRelInfo = std::make_unique<ExtraBoundCopyRelInfo>(
-        srcTableSchema, dstTableSchema, srcOffset, dstOffset, srcKey, dstKey);
-    auto boundCopyFromInfo = std::make_unique<BoundCopyFromInfo>(
-        tableSchema, std::move(boundFileScanInfo), containsSerial, std::move(extraCopyRelInfo));
+        srcTableSchema, dstTableSchema, srcNodeID, dstNodeID, srcKey, dstKey);
+    // Skip the first two columns.
+    expression_vector columnsToCopy{std::move(srcNodeID), std::move(dstNodeID), std::move(relID)};
+    for (auto i = 2; i < columnsToRead.size(); i++) {
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/binder/bind/bind_copy.cpp:178]: Added a constant `NUM_COLUMNS_TO_SKIP_IN_REL_FILE`, but kept inside the cpp file for now.

File: src/binder/bind/bind_copy.cpp [hunk 8]
```diff
@@ -269,19 +272,28 @@ expression_vector Binder::bindExpectedRelFileColumns(
...
         assert(dstTable->tableType == TableType::NODE);
-        readerConfig.columnTypes.push_back(
-            reinterpret_cast<NodeTableSchema*>(srcTable)->getPrimaryKey()->getDataType()->copy());
-        readerConfig.columnTypes.push_back(
-            reinterpret_cast<NodeTableSchema*>(dstTable)->getPrimaryKey()->getDataType()->copy());
-        columns.push_back(createVariable(srcColumnName, arrowColumnType));
-        columns.push_back(createVariable(dstColumnName, arrowColumnType));
+        auto srcPKColumnType =
+            reinterpret_cast<NodeTableSchema*>(srcTable)->getPrimaryKey()->getDataType()->copy();
+        if (srcPKColumnType->getLogicalTypeID() == LogicalTypeID::SERIAL) {
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/binder/bind/bind_copy.cpp:277]: Right. We should still read INT64 into SERIAL as its physical type is INT64. I will list this as a minor refactoring TODO.

File: src/include/processor/operator/index_lookup.h [hunk 9]
```diff
@@ -15,20 +15,21 @@ class PrimaryKeyIndex;
 namespace processor {
 
 struct IndexLookupInfo {
+    common::table_id_t tableID;
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/include/processor/operator/index_lookup.h:18]: Good catch. Deleted. This was used, but the usages were cleaned up later, and the variable remains still.

File: src/include/processor/operator/scan/scan_multi_node_tables.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+    bool getNextTuplesInternal(ExecutionContext* context) override;
+
+    inline std::unique_ptr<PhysicalOperator> clone() override {
+        return make_unique<ScanMultiNodeTables>(inVectorPos, outVectorsPos, tables,
+            tableIDToScanColumnIds, children[0]->clone(), id, paramsString);
+    }
+
+private:
+    std::unordered_map<common::table_id_t, storage::NodeTable*> tables;
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/include/processor/operator/scan/scan_multi_node_tables.h:28]: Right. Separate this into another PR against master.

File: src/include/processor/operator/scan/scan_rel_regular_columns.h [hunk 11]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include "processor/operator/filtering_operator.h"
+#include "processor/operator/scan/scan_rel_table.h"
+
+namespace kuzu {
+namespace processor {
+
+class ScanRelRegularColumns : public ScanRelTable, SelVectorOverWriter {
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/include/processor/operator/scan/scan_rel_regular_columns.h:9]: True. I'm keeping the same hierarchical structure as in master. This should be a separate issue.

File: src/processor/map/map_copy_from.cpp (hunk 1 above)
Comment (ray6080) [reply src/processor/map/map_copy_from.cpp:109]: See #2321

File: src/processor/operator/index_lookup.cpp [hunk 12]
```diff
@@ -35,113 +31,61 @@ std::unique_ptr<PhysicalOperator> IndexLookup::clone() {
...
+    auto resultVector = resultSet->getValueVector(info.resultVectorPos).get();
+    fillOffsetArraysFromVector(transaction, info, keyVector, resultVector);
+}
+
+void IndexLookup::checkNullKeys(ValueVector* keyVector) {
+    if (keyVector->hasNoNullsGuarantee()) {
+        return;
+    }
+    for (auto i = 0u; i < keyVector->state->selVector->selectedSize; i++) {
+        auto pos = keyVector->state->selVector->selectedPositions[i];
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/processor/operator/index_lookup.cpp:46]: Left TODOs for now.

File: src/processor/operator/index_lookup.cpp [hunk 13]
```diff
@@ -35,113 +31,61 @@ std::unique_ptr<PhysicalOperator> IndexLookup::clone() {
...
+    assert(resultVector->dataType.getPhysicalType() == PhysicalTypeID::INT64);
+    auto offsets = (offset_t*)resultVector->getData();
+    auto numKeys = keyVector->state->selVector->selectedSize;
     switch (info.pkDataType->getLogicalTypeID()) {
-    case LogicalTypeID::SERIAL: {
-        for (auto i = 0u; i < length; i++) {
-            offsets[i] = static_cast<offset_t>(dynamic_cast<arrow::Int64Array*>(array)->Value(i));
-        }
-    } break;
     case LogicalTypeID::INT64: {
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/processor/operator/index_lookup.cpp:59]: `SERIAL` is handled differently from `INT64`.

File: src/processor/map/map_copy_from.cpp (hunk 3 above)
Comment (ray6080) [reply src/processor/map/map_copy_from.cpp:89]: Left a TODO for now.

File: src/binder/bind/bind_copy.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ray6080) [reply src/binder/bind/bind_copy.cpp:135]: See issue #2331.

File: src/processor/operator/partitioner.cpp [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+void Partitioner::constructDataChunk(
+    DataChunk* dataChunk, const std::vector<DataPos>& dataPoses, ResultSet* resultSet) {
+    for (auto i = 0u; i < dataPoses.size(); i++) {
+        dataChunk->insert(i, resultSet->getValueVector(dataPoses[i]));
+    }
+}
+
+void Partitioner::copyDataToPartitions(ValueVector* partitionIdxes, DataChunk* chunkToCopyFrom,
+    PartitioningBuffer* partitioningBuffer, MemoryManager* memoryManager) {
+    for (auto i = 0u; i < chunkToCopyFrom->state->selVector->selectedSize; i++) {
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/processor/operator/partitioner.cpp:122]: Added todo

File: src/processor/operator/partitioner.cpp [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+        auto posToCopyFrom = chunkToCopyFrom->state->selVector->selectedPositions[i];
+        auto partitionIdx = partitionIdxes->getValue<partition_idx_t>(posToCopyFrom);
+        auto partition = partitioningBuffer->partitions[partitionIdx].get();
+        if (partition->chunks.empty() ||
+            partition->chunks.back()->state->selVector->selectedSize == DEFAULT_VECTOR_CAPACITY) {
+            auto newDataChunk = std::make_unique<DataChunk>(chunkToCopyFrom->getNumValueVectors());
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/processor/operator/partitioner.cpp:129]: Added TODO

**Pull request merged**